### ISSUE TITLE
Get the first deployment done

### ISSUE THREAD

**Issue by mlissner** [freelawproject/awscloud#3]
@tewen did really good work here and I've been remiss in actually using it. That needs to end and for that to happen, I need to deploy his cloudformation work to AWS. Let's see how it goes....

**Comment by mlissner** [comment]
The process here was pretty smooth. I already had sam installed, I didn't bother with the IDE stuff, and I mostly just ran the guided installation, which resulted in this:

```
↪ sam build --use-container
Starting Build inside a container
Building codeuri: /home/mlissner/Programming/awscloud/cl/recap_email runtime: python3.8 metadata: {} architecture: x86_64 functions: ['RecapEmailFunction']

Fetching public.ecr.aws/sam/build-python3.8:latest-x86_64 Docker container image..........................................................................................................................................................................................................................................................................................................................................................................................................................................
Mounting /home/mlissner/Programming/awscloud/cl/recap_email as /tmp/samcli/source:ro,delegated inside runtime container

Build Succeeded

Built Artifacts  : .aws-sam/build
Built Template   : .aws-sam/build/template.yaml

Commands you can use next
=========================
[*] Invoke Function: sam local invoke
[*] Test Function in the Cloud: sam sync --stack-name {stack-name} --watch
[*] Deploy: sam deploy --guided
        
Running PythonPipBuilder:ResolveDependencies
Running PythonPipBuilder:CopySource
(awscloud) 04:21:18::mlissner@gabbro::~/Programming/awscloud/cl 
↪ sam deploy --guided

Configuring SAM deploy
======================

	Looking for config file [samconfig.toml] :  Not found

	Setting default arguments for 'sam deploy'
	=========================================
	Stack Name [sam-app]: recap-dot-email
	AWS Region [us-east-1]: us-west-2
	#Shows you resources changes to be deployed and require a 'Y' to initiate deploy
	Confirm changes before deploy [y/N]: y
	#SAM needs permission to be able to create roles to connect to the resources in your template
	Allow SAM CLI IAM role creation [Y/n]: y
	#Preserves the state of previously provisioned resources when an operation fails
	Disable rollback [y/N]: 
	RecapEmailFunction may not have authorization defined, Is this okay? [y/N]: 
Error: Security Constraints Not Satisfied!
(awscloud) 04:31:58::mlissner@gabbro::~/Programming/awscloud/cl 
↪ sam deploy --guided

Configuring SAM deploy
======================

	Looking for config file [samconfig.toml] :  Not found

	Setting default arguments for 'sam deploy'
	=========================================
	Stack Name [sam-app]: recap-dot-email
	AWS Region [us-east-1]: us-west-2
	#Shows you resources changes to be deployed and require a 'Y' to initiate deploy
	Confirm changes before deploy [y/N]: y
	#SAM needs permission to be able to create roles to connect to the resources in your template
	Allow SAM CLI IAM role creation [Y/n]: y
	#Preserves the state of previously provisioned resources when an operation fails
	Disable rollback [y/N]: n
	RecapEmailFunction may not have authorization defined, Is this okay? [y/N]: y
	Save arguments to configuration file [Y/n]: y
	SAM configuration file [samconfig.toml]: 
	SAM configuration environment [default]: 

	Looking for resources needed for deployment:
	Creating the required resources...
	Successfully created!
	 Managed S3 bucket: aws-sam-cli-managed-default-samclisourcebucket-vc7oy5nsaht1
	 A different default S3 bucket can be set in samconfig.toml

	Saved arguments to config file
	Running 'sam deploy' for future deployments will use the parameters saved above.
	The above parameters can be changed by modifying samconfig.toml
	Learn more about samconfig.toml syntax at 
	https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-config.html

Uploading to recap-dot-email/5eb21e209010f18b072ca2ae762ece48  459150 / 459150  (100.00%)

	Deploying with following values
	===============================
	Stack name                   : recap-dot-email
	Region                       : us-west-2
	Confirm changeset            : True
	Disable rollback             : False
	Deployment s3 bucket         : aws-sam-cli-managed-default-samclisourcebucket-vc7oy5nsaht1
	Capabilities                 : ["CAPABILITY_IAM"]
	Parameter overrides          : {}
	Signing Profiles             : {}

Initiating deployment
=====================
Uploading to recap-dot-email/07cd54bc04ff37481fbe67c87aa350d2.template  1232 / 1232  (100.00%)

Waiting for changeset to be created..

CloudFormation stack changeset
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Operation                                      LogicalResourceId                              ResourceType                                   Replacement                                  
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+ Add                                          RecapEmailFunctionRecapEmailPermissionProd     AWS::Lambda::Permission                        N/A                                          
+ Add                                          RecapEmailFunctionRole                         AWS::IAM::Role                                 N/A                                          
+ Add                                          RecapEmailFunction                             AWS::Lambda::Function                          N/A                                          
+ Add                                          ServerlessRestApiDeployment1d6c579a39          AWS::ApiGateway::Deployment                    N/A                                          
+ Add                                          ServerlessRestApiProdStage                     AWS::ApiGateway::Stage                         N/A                                          
+ Add                                          ServerlessRestApi                              AWS::ApiGateway::RestApi                       N/A                                          
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Changeset created successfully. arn:aws:cloudformation:us-west-2:xxx:changeSet/samcli-deploy-xxx/xxx-66ab-41d1-8d50-xxx


Previewing CloudFormation changeset before deployment
======================================================
Deploy this changeset? [y/N]: y

2022-01-04 16:36:06 - Waiting for stack create/update to complete

CloudFormation events from stack operations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
ResourceStatus                                 ResourceType                                   LogicalResourceId                              ResourceStatusReason                         
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
CREATE_IN_PROGRESS                             AWS::IAM::Role                                 RecapEmailFunctionRole                         -                                            
CREATE_IN_PROGRESS                             AWS::IAM::Role                                 RecapEmailFunctionRole                         Resource creation Initiated                  
CREATE_COMPLETE                                AWS::IAM::Role                                 RecapEmailFunctionRole                         -                                            
CREATE_IN_PROGRESS                             AWS::Lambda::Function                          RecapEmailFunction                             -                                            
CREATE_IN_PROGRESS                             AWS::Lambda::Function                          RecapEmailFunction                             Resource creation Initiated                  
CREATE_COMPLETE                                AWS::Lambda::Function                          RecapEmailFunction                             -                                            
CREATE_IN_PROGRESS                             AWS::ApiGateway::RestApi                       ServerlessRestApi                              -                                            
CREATE_IN_PROGRESS                             AWS::ApiGateway::RestApi                       ServerlessRestApi                              Resource creation Initiated                  
CREATE_COMPLETE                                AWS::ApiGateway::RestApi                       ServerlessRestApi                              -                                            
CREATE_IN_PROGRESS                             AWS::Lambda::Permission                        RecapEmailFunctionRecapEmailPermissionProd     Resource creation Initiated                  
CREATE_IN_PROGRESS                             AWS::ApiGateway::Deployment                    ServerlessRestApiDeployment1d6c579a39          -                                            
CREATE_IN_PROGRESS                             AWS::Lambda::Permission                        RecapEmailFunctionRecapEmailPermissionProd     -                                            
CREATE_IN_PROGRESS                             AWS::ApiGateway::Deployment                    ServerlessRestApiDeployment1d6c579a39          Resource creation Initiated                  
CREATE_COMPLETE                                AWS::ApiGateway::Deployment                    ServerlessRestApiDeployment1d6c579a39          -                                            
CREATE_IN_PROGRESS                             AWS::ApiGateway::Stage                         ServerlessRestApiProdStage                     -                                            
CREATE_IN_PROGRESS                             AWS::ApiGateway::Stage                         ServerlessRestApiProdStage                     Resource creation Initiated                  
CREATE_COMPLETE                                AWS::ApiGateway::Stage                         ServerlessRestApiProdStage                     -                                            
CREATE_COMPLETE                                AWS::Lambda::Permission                        RecapEmailFunctionRecapEmailPermissionProd     -                                            
CREATE_COMPLETE                                AWS::CloudFormation::Stack                     recap-dot-email                                -                                            
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

CloudFormation outputs from deployed stack
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Outputs                                                                                                                                                                                    
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Key                 RecapEmailApi                                                                                                                                                          
Description         API Gateway endpoint URL for Prod stage for Free Look function                                                                                                         
Value               https://xxx.execute-api.us-west-2.amazonaws.com/Prod/recap-email/                                                                                               

Key                 RecapEmailFunction                                                                                                                                                     
Description         Free Look Email Lambda Function ARN                                                                                                                                    
Value               arn:aws:lambda:us-west-2:xxx:function:recap-dot-email-RecapEmailFunction-xxx                                                                         

Key                 RecapEmailFunctionIamRole                                                                                                                                              
Description         Implicit IAM Role created for the Free Look Email function                                                                                                             
Value               arn:aws:iam::xxx:role/recap-dot-email-RecapEmailFunctionRole-xxx                                                                                    
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Successfully created/updated stack - recap-dot-email in us-west-2
```

Sorta seems like it worked. @tewen, what next? I don't think it's sensitive, but I'll slack you the API endpoint. It comes returns an error, but that's sort of to be expected, I think.

**Comment by mlissner** [comment]
OK, @tewen sent me [this video][v] explaining that the next step was to create a recipient ruleset in AWS SES, so I did that. It went pretty well, but:

1. The console complained that I needed to verify domain ownership, but that was already done. 

2. When I tried to create the ruleset, it had a permissions error that I had to fix via [this link][aws]. That link told me to put my user account's information into the S3 buckets permissions, which was pretty weird, but when I did it, it worked. Hm.

[v]: https://threadit.app/thread/qu326d97drofbeljmiqr/message/ya9zed6rety64kn6lgar5vg2?utm_medium=referral-link
[aws]: https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-permissions.html

The last part I haven't yet done is activate the ruleset. When I start doing so, I get the warning that:

> Set default-rule-set as the active rule set for your account in US West (Oregon)? Only one receipt rule set can be active at a time.

I worried a bit that if I made this active, it would start applying to all of my emails, not only those sent to recap.email, but [the docs][d] seem to indicate that it'll be fine, so I enabled it. They say:

> If the recipient of an incoming email matches a recipient specified in the conditions for the receipt rule, then Amazon SES performs the actions specified in that receipt rule. 

Which implies that if the recipient doesn't match, then the usual stuff is done with it.

[d]: https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-receipt-rules.html
